### PR TITLE
`miette` integration for all errors in Core, Validator, and `cedar-policy`

### LINKED ISSUE #182: `miette` integration for all errors
### Category

User level API changes

### Describe the feature you'd like to request

In #148, we refactored some of our errors to use `miette`.  We should do this for all errors in Core and Validator.

### Describe the solution you'd like

As above.

### Describe alternatives you've considered

N/A

### Additional context

_No response_

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [X] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

Derives (or manually implements) `miette::Diagnostic` for all our error types.  Tries to make good use of the `help()` field, but does not yet assign error codes or propagate source locations; those can be done in followup PRs (and have their own GitHub issues).

No breaking changes to the public API.  Some error messages have changed, and in particular, in some cases, information that used to be in the `Display` for an error is now in other `miette` fields.  I have updated the CHANGELOG with a note about this.  I'm not sure where else we should document it.

To display a `miette::Diagnostic` in a pretty fashion, convert it to `miette::Report` using `From`/`Into`.  Note that `?` automatically converts errors, so it is sufficient to declare the error type of your function (or `main()`) as `miette::Report`.  Note also that you need the `fancy` feature of `miette` for prettiest output; we enable this in the CLI crate but not in our library crates, as recommended by `miette`. ("You should only [enable the `"fancy"` feature] in your toplevel crate, as the fancy feature pulls in a number of dependencies that libraries and such might not want.")

Sample outputs from the CLI:
image
image

## Issue #, if available

Resolves #182

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy-validator/src/err.rs [hunk 1]
```diff
@@ -22,35 +22,44 @@ use cedar_policy_core::{
...
+        "any entity types appearing anywhere in a schema need to be declared in `entityTypes`"
+    ))]
     UndeclaredEntityTypes(HashSet<String>),
     /// Undeclared action(s) used in the `memberOf` field of an action.
     #[error("undeclared action(s): {0:?}")]
+    #[diagnostic(help("any actions appearing in `memberOf` need to be declared in `actions`"))]
     UndeclaredActions(HashSet<String>),
     /// Undeclared common type(s) used in entity or context attributes.
     #[error("undeclared common type(s): {0:?}")]
+    #[diagnostic(help("any common types used in entity or context attributes need to be declared in `commonTypes`"))]
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/err.rs:62]: is this the error we get if we try to have common types reference other common types?

if so I can update the doc comment on the variant as well

File: cedar-policy-validator/src/type_error.rs [hunk 2]
```diff
@@ -44,6 +46,47 @@ pub struct TypeError {
     pub(crate) kind: TypeErrorKind,
 }
 
+// custom impl of Diagnostic: source location is from .source_span(),
```
Comment (andrewmwells-amazon) [inline cedar-policy-validator/src/type_error.rs:49]:
```suggestion
// custom impl of `Diagnostic`: source location is from .source_span(),
```

File: cedar-policy-core/src/evaluator/err.rs [hunk 3]
```diff
@@ -15,27 +15,60 @@
...
     advice: Option<String>,
 }
 
-impl Display for EvaluationError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        if let Some(help_msg) = &self.advice {
-            write!(f, "{}. {}", self.error_kind, help_msg)
-        } else {
-            self.error_kind.fmt(f)
+// custom impl of Diagnostic: non-trivial implementation of `help()`.
```
Comment (andrewmwells-amazon) [inline cedar-policy-core/src/evaluator/err.rs:33]:
```suggestion
// custom impl of `Diagnostic`: non-trivial implementation of `help()`,
```

File: cedar-policy/src/api.rs [hunk 4]
```diff
@@ -1527,6 +1572,43 @@ impl<'a> From<cedar_policy_validator::ValidationWarning<'a>> for ValidationWarni
     }
 }
 
+// custom impl of Diagnostic: source location is from .location, everything else
```
Comment (andrewmwells-amazon) [inline cedar-policy/src/api.rs:1575]:
```suggestion
// custom impl of `Diagnostic`: source location is from .location, everything else
```

File: cedar-policy-core/src/est/err.rs [hunk 5]
```diff
@@ -58,14 +64,15 @@ pub enum FromJsonError {
     },
     /// Error thrown while processing string escapes
     #[error("invalid escape patterns: {:?}", .0.iter().map(|e| e.to_string()).collect::<Vec<String>>())]
-    UnescapeError(Vec<unescape::UnescapeError>),
+    UnescapeError(#[related] Vec<unescape::UnescapeError>),
```
[the remark this answers is not in the document]
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/est/err.rs:67]: It doesn't bother me too much. Using the `{:?}` debug formatting in error messages doesn't always work out well, but if it formats nicely in this case it's fine

[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-core/src/est/err.rs]: Fixed this up some, including the error message in `unescape::UnescapeError` itself

File: cedar-policy-validator/src/err.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/err.rs]: Updated

**Review by john-h-kastner-aws — APPROVED** [review]

**Pull request merged**